go/obscuronode/host/host.go

```diff
@@ -538,30 +466,114 @@ func (a *Node) monitorBlocks() {
 
 	// only process blocks if the node is running
 	for atomic.LoadInt32(a.stopNodeInterrupt) == 0 {
-		latestBlkHeader := <-listener
+		blkHeader := <-listener
+
 		// don't process blocks if the node is stopping
 		if atomic.LoadInt32(a.stopNodeInterrupt) == 1 {
 			return
 		}
-		block, err := a.ethClient.BlockByHash(latestBlkHeader.Hash())
+
+		// ignore blocks if bootstrapping is happening
+		if atomic.LoadInt32(a.ignoreBootstrapping) == 0 {
+			log.Trace("Node in bootstrap - ignoring block %s", blkHeader.Hash())
+			continue
+		}
+
+		block, err := a.ethClient.BlockByHash(blkHeader.Hash())
 		if err != nil {
-			log.Panic("could not fetch block for hash %s. Cause: %s", latestBlkHeader.Hash().String(), err)
+			log.Panic("could not fetch block for hash %s. Cause: %s", blkHeader.Hash().String(), err)
 		}
 		blockParent, err := a.ethClient.BlockByHash(block.ParentHash())
 		if err != nil {
 			log.Panic("could not fetch block's parent with hash %s. Cause: %s", block.ParentHash().String(), err)
 		}
 
-		nodecommon.LogWithID(a.shortID, "Received a new block b_%d(%d)",
-			obscurocommon.ShortHash(latestBlkHeader.Hash()),
-			latestBlkHeader.Number.Uint64())
+		nodecommon.LogWithID(
+			a.shortID,
+			"Received a new block b_%d(%d)",
+			obscurocommon.ShortHash(blkHeader.Hash()),
+			blkHeader.Number.Uint64(),
+		)
 		a.blockRPCCh <- blockAndParent{obscurocommon.EncodeBlock(block), obscurocommon.EncodeBlock(blockParent)}
 	}
 }
 
-func (a *Node) awaitSecret() {
-	// start listening for l1 blocks that contain the response to the request
-	for {
+func (a *Node) bootstrapNode() types.Block {
+	var err error
+	var nextBlk *types.Block
+
+	// build up from the genesis block
+	// todo update to bootstrap from the last block in storage
+	// todo the genesis block should be the block where the contract was deployed
+	currentBlock, err := a.ethClient.BlockByNumber(big.NewInt(0))
+	if err != nil {
+		panic(err)
+	}
+
+	nodecommon.LogWithID(a.shortID, "Started node bootstrap with block %d", currentBlock.NumberU64())
+
+	for startTime := time.Now(); ; {
+		// TODO ingest one block at a time or batch the blocks
+		result := a.EnclaveClient.IngestBlocks([]*types.Block{currentBlock})
+		if !result[0].IngestedBlock && result[0].BlockNotIngestedCause != "" {
+			nodecommon.LogWithID(
+				a.shortID,
+				"Failed to ingest block b_%d. Cause: %s",
+				obscurocommon.ShortHash(result[0].BlockHeader.Hash()),
+				result[0].BlockNotIngestedCause,
+			)
+		}
+		a.storeBlockProcessingResult(result[0])
+
+		nextBlk, err = a.ethClient.BlockByNumber(big.NewInt(currentBlock.Number().Int64() + 1))
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) {
+				break
+			}
+			panic(err)
+		}
+		currentBlock = nextBlk
+
+		if time.Since(startTime)%(10*time.Second) == 0 {
```

Comment:
Are we sure this reliably works as intended? Looks a bit scary to me (like could it spam out multiple times during each 10th second, could it miss the 10th second and not reliably log, etc.)

For clarity I'd be tempted to do something like name the variable `lastProgressLog`, and then
``` 
if time.Since(lastProgressLog) > 10 * time.Second {
    // log the msg
    lastProgressLog = time.Now()
}
```

Reply:
That's a good concern. This should work, but I triple checked https://go.dev/play/p/MZHrPvKCVPp :)